overlord/servicestate/quota_handlers.go

```diff
@@ -123,54 +124,118 @@ func (m *ServiceManager) doQuotaControl(t *state.Task, _ *tomb.Tomb) error {
 		// modifications to state will be committed by the
-		// unlocking in restartSnapServices. If snapd gets
+		// unlocking at the end of this task. If snapd gets
 		// restarted before the end of this task, all the
 		// modifications would be redone, and those
 		// non-idempotent parts of the task would fail.
 		// For this reason we record together with the changes
 		// in state the fact that the changes were made,
 		// to avoid repeating them.
 		// What remains for this task handler is just to
-		// restart services which will happen regardless if we
-		// get rebooted after unlocking the state - if we got
-		// rebooted before unlocking the state, none of the
-		// changes we made to state would be persisted and we
-		// would run through everything above here again, but
-		// the second time around EnsureSnapServices would end
-		// up doing nothing since it is idempotent.  So in the
-		// rare case that snapd gets restarted but is not a
+		// refresh security profiles and restart services which
+		// will happen regardless if we get rebooted after
+		// unlocking the state - if we got rebooted before unlocking
+		// the state, none of the changes we made to state would
+		// be persisted and we would run through everything above
+		// here again, but the second time around EnsureSnapServices
+		// would end up doing nothing since it is idempotent.
+		// So in the rare case that snapd gets restarted but is not a
 		// reboot also record which services do need
 		// restarting. There is a small chance that services
 		// will be restarted again but is preferable to the
 		// quota not applying to them.
-		if err := rememberQuotaStateUpdated(t, appsToRestartBySnap); err != nil {
+		if err := rememberQuotaStateUpdated(t, servicesAffected, refreshProfiles); err != nil {
 			return err
 		}
-
 	}
 
-	if err := restartSnapServices(st, t, appsToRestartBySnap, perfTimings); err != nil {
-		return err
+	if len(servicesAffected) > 0 {
+		var prevTaskSet *state.TaskSet
+		chg := t.Change()
+		queueTasks := func(ts *state.TaskSet) {
+			if prevTaskSet != nil {
+				prevTaskSet.WaitAll(ts)
+			} else {
+				ts.WaitFor(t)
+			}
+			chg.AddAll(ts)
+			prevTaskSet = ts
+		}
+
+		if refreshProfiles {
+			queueTasks(addRefreshProfileTasks(st, servicesAffected))
+		}
+		queueTasks(addRestartServicesTasks(st, qc.QuotaName, servicesAffected))
 	}
+
 	t.SetStatus(state.DoneStatus)
 	return nil
 }
 
+func addRefreshProfileTasks(st *state.State, servicesAffected map[*snap.Info][]*snap.AppInfo) *state.TaskSet {
+	ts := state.NewTaskSet()
+	for info := range servicesAffected {
+		setupProfilesTask := st.NewTask("setup-profiles", fmt.Sprintf(i18n.G("Update snap %q (%s) security profiles"), info.SnapName(), info.Revision))
+		setupProfilesTask.Set("snap-setup", &snapstate.SnapSetup{
+			SideInfo: &snap.SideInfo{
+				RealName: info.SnapName(),
+				Revision: info.Revision,
+			},
+		})
+		setupProfilesTask.WaitAll(ts)
```

Comment:
So in theory they could run in parallel/unordered, the issue is it breaks all the tests due to switching order of syscalls/task ordering. I don't know how to fix that easily without doing some major changes to how we test.

Reply:
this is strange as it creates much more waiting that we really need. I think we fixed some code that would have made this terrible perfomance wise but is still not great. So we still want to chain them one by one at least here

Reply:
Your comment doesn't matter because we can't execute more than a setup-profiles at a time anyway. See the AddBlocked logic in ifacemgr.go

Reply:
Ah ok, didn't know this. I have updated the solution so we are back to a more linear approach.